### PR TITLE
ZCash transparent keyring

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/32156
ZCash uses [p2pkh](https://reference.cash/protocol/blockchain/addresses) Bitcoin address format with prefixes defined [here](https://github.com/zcash/librustzcash/blob/1e39daf6f83c4b62d645e74385d24299dc4414c7/components/zcash_address/src/kind/p2pkh.rs)
HD derivation path is m/44'/133'/ for the mainnet and m/44'/1' for the testnet

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by darkdh** [comment]
Please do link spec and summary, there is nothing useful in either issue or PR description.

File: components/brave_wallet/browser/keyring_service.cc [hunk 1]
```diff
@@ -632,9 +642,10 @@ void KeyringService::MigrateDerivedAccountIndex(PrefService* profile_prefs) {
...
   const std::vector<mojom::KeyringId> keyrings = {
-      mojom::KeyringId::kDefault,   mojom::KeyringId::kSolana,
-      mojom::KeyringId::kFilecoin,  mojom::KeyringId::kFilecoinTestnet,
-      mojom::KeyringId::kBitcoin84, mojom::KeyringId::kBitcoin84Testnet};
+      mojom::KeyringId::kDefault,      mojom::KeyringId::kSolana,
+      mojom::KeyringId::kFilecoin,     mojom::KeyringId::kFilecoinTestnet,
+      mojom::KeyringId::kBitcoin84,    mojom::KeyringId::kBitcoin84Testnet,
+      mojom::KeyringId::kZCashMainnet, mojom::KeyringId::kZCashTestnet};
 
   for (auto keyring_id : keyrings) {
```
Comment (supermassive) [inline components/brave_wallet/browser/keyring_service.cc:650]: No zcash here. 
There is no zcash keyrings stored in prefs which need to migrate storage format.

File: components/brave_wallet/browser/zcash/zcash_keyring.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  absl::optional<std::string> GetAddress(const mojom::BitcoinKeyId& key_id);
+
+  absl::optional<std::vector<uint8_t>> GetPubkey(
+      const mojom::BitcoinKeyId& key_id);
+
+  absl::optional<std::vector<uint8_t>> SignMessage(
+      const mojom::BitcoinKeyId& key_id,
+      base::span<const uint8_t, 32> message);
+
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_keyring.h:33]: why BitcoinKeyId for zcash?

File: components/brave_wallet/common/brave_wallet_types.h [hunk 3]
```diff
@@ -24,6 +24,8 @@ constexpr KeyringId kFilecoinKeyringId = KeyringId::kFilecoin;
 constexpr KeyringId kFilecoinTestnetKeyringId = KeyringId::kFilecoinTestnet;
 constexpr KeyringId kBitcoinKeyring84Id = KeyringId::kBitcoin84;
 constexpr KeyringId kBitcoinKeyring84TestId = KeyringId::kBitcoin84Testnet;
+constexpr KeyringId kZCashMainnetKeyringId = KeyringId::kZCashMainnet;
+constexpr KeyringId kZCashTestnetKeyringId = KeyringId::kZCashTestnet;
 }  // namespace mojom
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet_types.h:29]: Let's not do this. These are legacy keyring aliases and to be removed

File: components/brave_wallet/common/solana_utils_unittest.cc [hunk 4]
```diff
@@ -12,6 +12,7 @@
 #include "base/test/gtest_util.h"
 #include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
 #include "brave/components/brave_wallet/common/brave_wallet_constants.h"
+#include "brave/components/brave_wallet/common/encoding_utils.h"
```
Comment (supermassive) [inline components/brave_wallet/common/solana_utils_unittest.cc:15]: Let's move corresponging tests to `encoding_utils_unittest.h`

File: components/brave_wallet/common/zcash_utils.cc [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+std::string PubkeyToTransparentAddress(const std::vector<uint8_t>& pubkey,
+                                       bool testnet) {
+  std::vector<uint8_t> result = testnet ? std::vector<uint8_t>({0x1d, 0x25})
+                                        : std::vector<uint8_t>({0x1c, 0xb8});
+
+  std::vector<uint8_t> data_part = Hash160(pubkey);
+  result.insert(result.end(), data_part.begin(), data_part.end());
+  return Base58EncodeWithCheck(result);
+}
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.cc:21]: Needs unit test coverage

File: components/brave_wallet/common/zcash_utils.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_COMMON_ZCASH_UTILS_H_
+#define BRAVE_COMPONENTS_BRAVE_WALLET_COMMON_ZCASH_UTILS_H_
+
+#include <map>
```
Comment (supermassive) [inline components/brave_wallet/common/zcash_utils.h:9]: `<map>` is not needed

File: components/brave_wallet/browser/internal/hd_key.cc [hunk 7]
```diff
@@ -444,6 +445,10 @@ std::vector<uint8_t> HDKey::GetUncompressedPublicKey() const {
   return public_key;
 }
 
+std::string HDKey::GetZCashTransparentAddress(bool testnet) {
+  return PubkeyToTransparentAddress(public_key_, testnet);
+}
```
Comment (supermassive) [inline components/brave_wallet/browser/internal/hd_key.cc:450]: Unit test this please

**Review by nuo-xu — APPROVED** [review]
iOS++

**Review by supermassive — APPROVED** [review]

**Pull request merged**